### DISCUSSION TITLE
Field annotation inheritance - a use case for `Config.fields`

### DISCUSSION THREAD

**Discussion by nzig** [pydantic/pydantic#4242] [Pydantic V2]
Thanks for posting the plan for pydantic V2, and for taking comments on it. I appreciate the open approach.

I saw that you plan to deprecate `Config.fields`. I've actually been using it as a way to preserve field annotations with inheritance, like the following:
```python
from enum import Enum
from pydantic import BaseModel


class LaxModel(BaseModel):
    s: str

    class Config:
        fields = {"s": {"description": "This field should be an Animal name"}}
        
class Animal(str, Enum):
    cat = 'cat'
    dog = 'dog'
    
class StrictModel(LaxModel):
    s: Animal

from pprint import pprint
pprint(StrictModel.schema())
```
which prints
```
{'definitions': {'Animal': {'description': 'An enumeration.',
                            'enum': ['cat', 'dog'],
                            'title': 'Animal',
                            'type': 'string'}},
 'properties': {'s': {'allOf': [{'$ref': '#/definitions/Animal'}],
                      'description': 'This field should be an Animal name'}},
 'required': ['s'],
 'title': 'StrictModel',
 'type': 'object'}
```

If instead I had defined 
```python
class LaxModel(BaseModel):
    s: str = Field(description="This field should be an Animal name")
```
Then the `s: Animal` in the subclass completely overrides any annotations from the base class, and the only way to preserve it is to repeat all the parameters to `Field` in the type declaration in the subclass.

This might be an unintended side-effect of how config merging works, but I've found it useful to define models like this. I use strict models like this to restrict the allowed parameters a user can pass in a FastAPI handler, while allowing other callers of the same function the handler uses to pass a more extensive set of parameters.

I would love to hear your thoughts on how such field inheritance can be supported in pydantic V2, either through keeping `Config.fields` or through another mechanism.

**Comment by samuelcolvin** [comment]
Humm, I think this is basically a useful bug, [xkcd#1172](https://m.xkcd.com/1172/) (exaggerating obviously).

Config merging is currently a nightmare and simplifying it by getting rid of `config.fields` would be great (both for development and understandability).

I think the best approach in future would be to provide a method customise schema in which you could set description on fields based on parent classes/MRO.

**Reply by nzig** [reply]
I agree it's not elegant, and I'm fine with getting rid of it for V2.

I do think that inheriting `Field` attributes would be useful for more than just the description. Maybe something like
```python
class ModelWithOptionalField(BaseModel):
    s: Optional[str] = Field(min_length=1)


class Model(ModelWithOptionalField):
    s: str = Field("foo", merge=True)
```
could mean that `s` inherits any arguments that were passed `Field` in the base class that are not explicitly overridden? So that `Model.s` retained the `min_length=1`.

**Reply by samuelcolvin** [reply]
I could get behind a `merge` or `inherit` argument to `Field` :+1:.

**Comment by einarwar** [comment]
Hello. We are in the process of migrating to pydantic V2, and define a dictionary with descriptions that is added to the model-config of the base-class, and in the same way as @nzig is inherited by subclasses.

I have not been able to find anything in the documentation about this feature. What is the current status of `merge`/`inherit` as an argument to `Field`?

**Comment by ottowayi** [comment]
This is also an issue we've encountered in our attempt to upgrade to v2 as well.  We heavily rely on inheritance for field metadata and I'm having trouble finding the correct approach in v2.  Basically, we have a bunch of different modules for connecting to different types of devices, so we have some base models for the common config options and specify the metadata, then in the subclasses override (or add) the default values based the device or protocol.

```py

class DeviceConfig(BaseModel):
    log_verbosity: Annotated[int, Field(description='...', ge=1, le=5)]
    
class ConnectionConfig(BaseModel):
    host: Annotated[str, Field(alias='ansible_host', description='...')]
    port: Annotated[int, Field(alias='ansible_port', description='...', ge=1, le=65535)]

class SSHConfig(ConnectionConfig):
    port = 22

class TelnetConfig(ConnectionConfig):
    port = 23


class MyDeviceConfig(DeviceConfig, SSHConfig):
    ...
```

We want to be able to override/add defaults without having to re-specify all of the field metadata, but this is now an error and ensuring that field metadata is consistent over dozens of different models is going to be difficult.  I thought about specifying the `Fields` separately and re-using them:

```py

class FieldDefs:
    port = Field(alias='ansible_port', description='...', ge=1, le=65535)
    
class SSHConfig(ConnectionConfig):
    port: Annotated[int, FieldDefs.port] = 22
```

This would work for most things, but would still rely on re-defining the `Field` if a subclass requires some changes to any of the metadata.  A merge/inherit option would be ideal though.  Is that something that is in the works?

**Reply by wyattcolyn** [reply]
This problem is very similar in spirit to a problem I encountered when migrating to v2 https://github.com/pydantic/pydantic/discussions/7248
In your case it has to do with overriding/adding default values to the Field in the subclasses. In my case, I want to simply change the name of the field. 
However, in both cases the intent is to not have to re-specify the field metadata. Like you point out it is possible to define the Fields we want to use separately, but this is not ideal as this would still rely on redefining the Field. On second thought, I don't think this would solve my problem since an 'alias' is not the same as redefining the name itself. 
Sorry this isn't a solution, but I want to point out the pattern that is arising from the removal of the 'fields' setting from the Config class.
